src/NServiceBus.Core/Recoverability/FirstLevelRetries/FailureInfoStorage.cs

```diff
@@ -0,0 +1,134 @@
+        {
+            lock (lockObject)
+            {
+                FailureInfoNode node;
+                if (failureInfoPerMessage.TryGetValue(messageId, out node))
+                {
+                    // We have seen this message before, just update the counter and store exception.
+                    node.FailureInfo = new ProcessingFailureInfo(node.FailureInfo.NumberOfFailedAttempts + 1, exception);
+
+                    // Maintain invariant: leastRecentlyUsedMessages.First contains the LRU item.
+                    leastRecentlyUsedMessages.Remove(node.LeastRecentlyUsedEntry);
+                    leastRecentlyUsedMessages.AddLast(node.LeastRecentlyUsedEntry);
+                }
+                else
+                {
+                    if (failureInfoPerMessage.Count == maxElements)
+                    {
+                        // We have reached the maximum allowed capacity. Remove the LRU item.
+                        var leastRecentlyUsedEntry = leastRecentlyUsedMessages.First;
+                        failureInfoPerMessage.Remove(leastRecentlyUsedEntry.Value);
+                        leastRecentlyUsedMessages.RemoveFirst();
+                    }
+
+                    var newNode = new FailureInfoNode(
+                        messageId,
+                        new ProcessingFailureInfo(1, exception));
+
+                    failureInfoPerMessage[messageId] = newNode;
+
+                    // Maintain invariant: leastRecentlyUsedMessages.First contains the LRU item.
+                    leastRecentlyUsedMessages.AddLast(newNode.LeastRecentlyUsedEntry);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Retrieves message processing failure information for a given <paramref name="messageId"/>.
+        /// </summary>
+        /// <param name="messageId">Unique message identifier.</param>
+        /// <returns>
+        /// Returns null if no message processing failure information is stored for <paramref name="messageId"/>.
+        /// </returns>
+        public ProcessingFailureInfo GetFailureInfoForMessage(string messageId)
+        {
+            lock (lockObject)
+            {
+                FailureInfoNode node;
+                failureInfoPerMessage.TryGetValue(messageId, out node);
+                return node?.FailureInfo;
+            }
+        }
+
+        /// <summary>
+        /// Removes message processing failure information for a given <paramref name="messageId"/>.
+        /// </summary>
+        /// <param name="messageId">Unique message identifier.</param>
+        public void ClearFailureInfoForMessage(string messageId)
+        {
+            lock (lockObject)
+            {
+                failureInfoPerMessage.Remove(messageId);
+                leastRecentlyUsedMessages.Remove(messageId);
+            }
+        }
+
+        /// <summary>
+        /// Removes all items from the storage. 
+        /// </summary>
```

Comment:
This comment is just noise to me, remove? (see generic comment above regarding api doco for internal classes)